### DISCUSSION TITLE
Model Discriminator

### DISCUSSION THREAD

**Discussion by kempha** [laravel/framework#51521] [Ideas]
Add the ability to store several models in one table, distinguishing them by attribute, which will be stored as an auxiliary field (for example, attribute 'type'). The idea has already been implemented in the Doctrine ORM for the symfony (https://www.doctrine-project.org/projects/doctrine-orm/en/3.1/reference/inheritance-mapping.html#single-table-inheritance).

**Comment by ghost** [comment]
I think this is doable as it is now without attributes.

```
class Person extends Model
{
    protected $table = 'persons';
}


class Employee extends Person
{
    public const DISCRIMINATOR = 'something';

    /**
     * Mutator. Do not remove!
     */
    public function setDistriminatorAttribute(string $value): void
    {
        $this->attributes['distriminator'] = static::DISCRIMINATOR;
    }

    /**
     * Begin querying the model.
     *
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public static function query()
    {
        return parent::newQuery()->where('distriminator', static::DISCRIMINATOR);
    }
}
```

**Reply by kempha** [reply]
Yes, it works. But if you run the code: 
`Person:all();`
The result will be like this:
`[Person, Person, Person, ...]`
If you do the same thing in Symfony, then depending on the “discriminator” attribute, it will have its own required model. That is, as a result we get:
`[Person, Employee, Director, ...]`

**Reply by ghost** [reply]
you can also do that with a collection map if you overwrite the 

    /**
     * Create a new Eloquent Collection instance.
     *
     * @param  array  $models
     * @return \Illuminate\Database\Eloquent\Collection
     */
    public function newCollection(array $models = [])
    {
        return new Collection($models);
    }

in the model.
My personal opinion is that using attributes should be avoided unless that is the only option.

**Comment by henzeb** [comment]
It is ofcourse doable by yourself, or use a package. Google came up with https://github.com/tighten/parental
